core.ts

```diff
@@ -0,0 +1,1041 @@
+const noop = () => {}
+const deprecated: {
+	[ index: string ]: () => void
+} = {}
+const addDeprecated = (name: string, message: string) => {
+	deprecated[name] = () => {
+		console.warn('pinejs-client deprecated:', message)
+		deprecated[name] = noop
+	}
+}
+addDeprecated(
+	'expandFilter',
+	'`$filter: a: b: ...` is deprecated, please use `$filter: a: $any: { $alias: "x", $expr: x: b: ... }` instead.'
+)
+
+function defaults <T>(a: T | undefined, b: T | undefined, z: T): T
+function defaults <T>(a: T | undefined, z: T): T
+function defaults <T>(...args: (T | undefined)[]): T | undefined {
+	for (const arg of args) {
+		if (arg != null) {
+			return arg
+		}
+	}
+	return
+}
+const map = <T, R>(arr: T[], fn: (value: T) => R): R[] => {
+	const results = []
+	for (const key of arr) {
+		results.push(fn(key))
+	}
+	return results
```

Comment:
NIT: Do we need to define our own map function here? `Array.prototype.map` has very wide support
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map#Browser_compatibility

Reply:
It would still be a breaking change to do it, and it's kinda nice to have pretty near global compatibility..  that said this is a major version bump and we're removing a bunch of cruft so I think it makes sense (and you could still use 4.x or a polyfill for older versions) so I think the switch makes sense to do now

Reply:
Changed